Source/Core/DolphinQt/Debugger/CodeTraceDialog.cpp

```diff
@@ -0,0 +1,773 @@
+
+  boxes_layout->addWidget(m_backtrace);
+  boxes_layout->addWidget(m_verbose);
+  boxes_layout->addWidget(m_clear_on_loop);
+  boxes_layout->addWidget(m_change_range);
+  boxes_layout->addWidget(m_reprocess);
+
+  boxes_layout->addWidget(m_recorded_size);
+  boxes_layout->addWidget(m_record_limit_input);
+  boxes_layout->addWidget(m_results_size);
+  boxes_layout->addWidget(m_result_limit_input);
+  boxes_layout->addItem(new QSpacerItem(0, 0, QSizePolicy::Expanding, QSizePolicy::Maximum));
+  boxes_layout->addWidget(m_run_trace);
+
+  m_output_list = new QListWidget();
+  m_output_list->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+
+  QFont font(QStringLiteral("Monospace"));
+  font.setStyleHint(QFont::TypeWriter);
+  m_output_list->setFont(font);
+  m_output_list->setContextMenuPolicy(Qt::CustomContextMenu);
+
+  auto* layout = new QVBoxLayout();
+  layout->addLayout(input_layout);
+  layout->addLayout(boxes_layout);
+  layout->addWidget(m_output_list);
+
+  InfoDisp();
+
+  setLayout(layout);
+}
+
+void CodeTraceDialog::ConnectWidgets()
+{
+  connect(m_parent, &CodeWidget::BreakpointsChanged, this, &CodeTraceDialog::UpdateBreakpoints);
+  connect(m_run_trace, &QPushButton::toggled, this, &CodeTraceDialog::OnRunTrace);
+  connect(m_reprocess, &QPushButton::pressed, this, &CodeTraceDialog::DisplayTrace);
+  connect(m_change_range, &QCheckBox::toggled, [=](bool checked) {
+    if (checked)
+      OnChangeRange();
+  });
+  connect(m_output_list, &QListWidget::itemClicked, m_parent, [this](QListWidgetItem* item) {
+    m_parent->SetAddress(item->data(ADDRESS_ROLE).toUInt(),
+                         CodeViewWidget::SetAddressUpdate::WithUpdate);
+  });
+  connect(m_output_list, &CodeTraceDialog::customContextMenuRequested, this,
+          &CodeTraceDialog::OnContextMenu);
+}
+
+void CodeTraceDialog::ClearAll()
+{
+  std::vector<CodeTrace>().swap(m_code_trace);
+  std::vector<TraceOutput>().swap(m_trace_out);
+  m_reg.clear();
+  m_mem.clear();
+  m_output_list->clear();
+  m_bp1->setDisabled(true);
+  // i18n: Here, PC is an acronym for program counter, not personal computer.
+  m_bp1->setCurrentText(tr("Uses PC as trace starting point."));
+  m_bp2->setEnabled(true);
+  m_change_range->setChecked(false);
+  m_change_range->setDisabled(true);
+  UpdateBreakpoints();
+  InfoDisp();
+}
+
+void CodeTraceDialog::OnRunTrace(bool checked)
+{
+  if (!checked)
+  {
+    m_run_trace->setText(tr("Run Trace"));
+    ClearAll();
+    return;
+  }
+
+  m_record_limit = m_record_limit_input->value();
+  m_run_trace->setText(tr("Reset All"));
+  m_code_trace.clear();
+  m_code_trace.reserve(m_record_limit);
+
+  if (!CPU::IsStepping())
+    return;
+
+  CPU::PauseAndLock(true, false);
+  PowerPC::breakpoints.ClearAllTemporary();
+
+  // Keep stepping until the end_bp or timeout after five(?) seconds
+  using clock = std::chrono::steady_clock;
+  clock::time_point timeout = clock::now() + std::chrono::seconds(10);
```

Comment:
There's enough room for it I think.

Reply:
I'm not sure if we need it anymore. After changing to a simpler struct it can record 200,000 instructions in under 3 seconds. The question becomes how many instructions do we realistically need. In 5 seconds it can do 550,000 instructions, which uses about 20MB of storage. Right now it's limited to 200,000 instructions.